codegen/masm/src/codegen.rs

```diff
@@ -0,0 +1,657 @@
+use crate::constants::{AUX_TRACE, MAIN_TRACE};
+use crate::visitor::{
+    walk_boundary_constraints, walk_constant_bindings, walk_integrity_constraint_degrees,
+    walk_integrity_constraints, walk_periodic_columns, walk_public_inputs, AirVisitor,
+};
+use ir::{
+    constraints::{ConstraintRoot, Operation},
+    AccessType, AirIR, ConstantBinding, ConstantValueExpr, Identifier, IntegrityConstraintDegree,
+    NodeIndex, PeriodicColumn, PublicInput, TraceAccess, Value,
+};
+use std::collections::btree_map::{BTreeMap, Entry};
+
+use processor::math::{Felt, StarkField};
+use winter_prover::math::fft;
+
+use crate::config::CodegenConfig;
+use crate::writer::Writer;
+
+pub struct Codegen<'ast> {
+    /// Miden Assembly writer.
+    ///
+    /// Track indentation level, and performs basic validations for generated instructions and
+    /// closing of blocks.
+    writer: Writer,
+
+    /// Counts how many periodic columns have been visited so far.
+    ///
+    /// Periodic columns are visited in order, and the counter is the same as the columns ID.
+    periodic_column: u64,
+
+    /// Counts how many transition constraint roots have been visited so far.
+    transition_contraint: usize,
+
+    /// Map of the constants found while visitint the [AirIR].
+    ///
+    /// These values are later used to emit immediate values.
+    constants: BTreeMap<&'ast Identifier, &'ast ConstantValueExpr>,
+
+    /// The [AirIR] to visit.
+    ir: &'ast AirIR,
+
+    /// Configuration for the codegen.
+    config: CodegenConfig,
+}
+
+/// Converts a column ID to an element position.
+///
+/// The element position is then used to determine the memory location to load, and which elements
+/// of the word must be kept.
+///
+/// The position is used for both the cached value of `z^exp`, which is the point at which the
+/// polynomial equivalent to the periodic column is evaluated, and the result of evaluating the
+/// periodic polynomial at that point. Both values are stored in a unique memory location, the
+/// memomry is expected to be $[0, 0, q_1, q_0]$, this uses additional memory but simplifies code
+/// generation and requires fewer cycles to compute the result, because packing values is not
```

Comment:
> Is (q0, q1) evaluations of the periodic column polynomial at z^exp?

That is the format in memory for the `z^exp` and for the evaluation. Basically I'm trying to say there is one element per memory location.

> Are we storing the evaluations in the lower part of the word so that they are at the top of the stack when we load them from memory? If so, shouldn't it be [0, 0, q_0, q_1]?

I documented stack order, so the order in there is how the code sees it after the load.

> The text says that we are storing both z^exp and the evaluations in the memory address. So, would the contents of memory address look like [zexp_0, zexp_1, q_0, q_1]?

Both are store in memory, but in different memory ranges. Both are stored as `[0,0,q_1,q_0]`

Reply:
I guess what confused me here is mentioning of `z^exp`. If it was something like "Values of periodic columns are stored at distinct memory addresses such that each value occupies a single memory word with the two most significant word elements set to zeros (i.e., [q0, q1, 0, 0])" - I would have understood it.

Reply:
updated